### PR TITLE
fix: ensure consistency in method signatures across destination connectors

### PR DESCRIPTION
**Pull request by rbiseck3**
### Description
* Make sure all destination connectors implement the base abstract methods using the same signatures. 
* Also leverage conform dict in the base methods to make sure it's called in a consistent fashion.
* Additional updates to move the common code into the base destination connector class

### CODE REVIEW

**Comment by rbiseck3** [comment]
@ahmetmeleq I believe this should handle the issue you brought up with `conform_dict()`

**Comment by ryannikolaidis** [comment]
I think you're going to hit issues with the unit tests leveraging assignment like [here](https://github.com/Unstructured-IO/unstructured/blob/e2f0de3c5049850be191e140cdc28af811ce907e/test_unstructured_ingest/unit/test_sql_conform_dict.py#L72). should be quick bump

**Comment by ryannikolaidis** [comment]
bump PR title (for conventional) and description to include some more info about the what and why? also maybe include a note in the changelog?

**Comment by ryannikolaidis** [comment]
not sure if this is in scope for this PR, but calling `write_elements()` rather than `write()` will still hit issues with some of these connectors, right? i.e. with Pinecone we [do formatting ](https://github.com/Unstructured-IO/unstructured/blob/e2f0de3c5049850be191e140cdc28af811ce907e/unstructured/ingest/connector/pinecone.py#L111-L139)directly in `write()`

File: unstructured/ingest/connector/azure_cognitive_search.py [hunk 1]
```diff
@@ -150,8 +150,6 @@ def write(self, docs: t.List[BaseSingleIngestDoc]) -> None:
             local_path = doc._output_filename
             with open(local_path) as json_file:
                 json_content = json.load(json_file)
-                for content in json_content:
-                    self.conform_dict(data=content)
```
Comment (ahmetmeleq) [inline unstructured/ingest/connector/azure_cognitive_search.py:154]: How will ingest conform the dicts now since we remove it from write methods?

**Comment by ahmetmeleq** [comment]
> not sure if this is in scope for this PR, but calling `write_elements()` rather than `write()` will still hit issues with some of these connectors, right? i.e. with Pinecone we [do formatting ](https://github.com/Unstructured-IO/unstructured/blob/e2f0de3c5049850be191e140cdc28af811ce907e/unstructured/ingest/connector/pinecone.py#L111-L139)directly in `write()`

I think this PR should: 
1. isolate all formatting logic from `write` methods
2. call the isolated formatting logic both in `write` , and `write_elements` to support both how ingest and other downstream applications

File: test_unstructured_ingest/dest/azure-cognitive-search.sh [hunk 2]
```diff
@@ -21,8 +19,8 @@ DESTINATION_INDEX="utic-test-ingest-fixtures-output-$random_id"
 # 2023-07-01-Preview, 2021-04-30-Preview, 2020-06-30-Preview
 API_VERSION=2023-07-01-Preview
 
-if [ -z "$AZURE_SEARCH_API_KEY" ]; then
-  echo "Skipping Azure Cognitive Search ingest test because AZURE_SEARCH_API_KEY env var is not set."
+if [ -z "$AZURE_SEARCH_API_KEY" ] || [ -z "$AZURE_SEARCH_ENDPOINT" ]; then
```
Comment (ahmetmeleq) [inline test_unstructured_ingest/dest/azure-cognitive-search.sh:22]: Why is this changed?

File: unstructured/ingest/connector/azure_cognitive_search.py (hunk 1 above)
Comment (rbiseck3) [reply unstructured/ingest/connector/azure_cognitive_search.py:154]: This was moved to the base classes in the destination connector, meaning whenever you call `write()` or `write_elements` not, it'll call the `conform_dict` method on your behalf.

File: test_unstructured_ingest/dest/azure-cognitive-search.sh (hunk 2 above)
Comment (rbiseck3) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:22]: Not sure how this was passing in CI, but that endpoint was a configurable value. It broke for me locally without this change.

File: unstructured/ingest/interfaces.py [hunk 3]
```diff
@@ -654,22 +654,34 @@ def __init__(self, write_config: WriteConfig, connector_config: BaseConnectorCon
...
+
+    @abstractmethod
+    def get_elements_dict(self, docs: t.List[BaseSingleIngestDoc]) -> t.List[t.Dict[str, t.Any]]:
         pass
 
     @abstractmethod
     def write_dict(self, *args, elements_dict: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
         pass
 
+    def write_raw_dict(self, *args, elements_dict: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
```
Comment (ahmetmeleq) [inline unstructured/ingest/interfaces.py:677]: I'd rather name this as `write_conformed_dict`, raw sounds as not processed / formatted

File: unstructured/ingest/connector/pinecone.py [hunk 4]
```diff
@@ -87,26 +87,28 @@ def upsert_batch(self, batch):
...
-                pool.map(self.upsert_batch, list(chunk_generator(dict_list, pinecone_batch_size)))
+                pool.map(
+                    self.upsert_batch, list(chunk_generator(elements_dict, pinecone_batch_size))
+                )
 
-    def write(self, docs: t.List[BaseIngestDoc]) -> None:
+    def get_elements_dict(self, docs: t.List[BaseSingleIngestDoc]) -> t.List[t.Dict[str, t.Any]]:
         dict_list: t.List[t.Dict[str, t.Any]] = []
         for doc in docs:
             local_path = doc._output_filename
```
Comment (ahmetmeleq) [inline unstructured/ingest/connector/pinecone.py]: Lines 119-138 are still not being called by `write_elements`, and they are essential while writing content. We need to isolate those parts as well. Both `write` and `write_elements` should work

**Comment by rbiseck3** [comment]
@ahmetmeleq, look at the latest changes. I created two base classes in the `BaseDestinationConnector` class:
* `conform_dict`: Changes that are needed in place
* `normalize_dict`: map to a new dict to accommodate the input needed by the destination 

These are then called in the base `write_raw_dict` method which gets called by the `write` and `write_elements` methods.

File: test_unstructured_ingest/dest/azure-cognitive-search.sh (hunk 2 above)
Comment (ahmetmeleq) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:22]: https://github.com/Unstructured-IO/unstructured/blob/a8a103bc5c378900c3c17bffde26a3b41287ff89/test_unstructured_ingest/dest/azure-cognitive-search.sh#L14C1-L14C22

The variable was defined at the top (being removed here) and it is not a secret value. We've made a similar discussion in a past PR, we don't want to bloat CI values with non secret variables

Comment (ahmetmeleq) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:22]: Also, the azure cognitive search credentials has recently been modified. If you have a local credential which is not updated, that might be the reason why it failed in local

**Comment by ahmetmeleq** [comment]
The design looks good 👍 , waiting for the tests - also two comments on naming and azure cognitive search test script

File: test_unstructured_ingest/dest/azure-cognitive-search.sh (hunk 2 above)
Comment (rbiseck3) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:22]: I assumed because we had that value in LastPass, it was being used in CI.

Comment (ryannikolaidis) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:24]: bump - can we remove? there should never be a case where this isn't defined given L14 right?

File: unstructured/ingest/interfaces.py (hunk 3 above)
Comment (ryannikolaidis) [reply unstructured/ingest/interfaces.py]: @rbiseck3 I do see your point, but I think I agree with @ahmetmeleq here. at a glance I immediately know what to expect from `normalize_and_write_dict` whereas I might not have assumed `write_raw_dict` does this without looking at the code.

Comment (ryannikolaidis) [reply unstructured/ingest/interfaces.py]: ahh, looks like you just bumped

File: unstructured/ingest/interfaces.py [hunk 5]
```diff
@@ -654,22 +655,54 @@ def __init__(self, write_config: WriteConfig, connector_config: BaseConnectorCon
...
+                dict_list.extend(dict_content)
+        return dict_list
 
     @abstractmethod
     def write_dict(self, *args, elements_dict: t.List[t.Dict[str, t.Any]], **kwargs) -> None:
         pass
 
+    def modify_and_write_dict(
+        self, *args, elements_dict: t.List[t.Dict[str, t.Any]], **kwargs
+    ) -> None:
```
Comment (ryannikolaidis) [inline unstructured/ingest/interfaces.py:697]: nit: maybe just a quick line docstring... modify=conform+normalize and why it's needed.

File: test_unstructured_ingest/dest/azure-cognitive-search.sh (hunk 2 above)
Comment (rbiseck3) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:24]: removed

Comment (ahmetmeleq) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:24]: Agreed that it shouldn't exist in Lastpass 👍 I tried updating but having permission issues

image

Comment (ahmetmeleq) [reply test_unstructured_ingest/dest/azure-cognitive-search.sh:24]: Updated now 👍

**Comment by potter-potter** [comment]
This is great! `conform_dict` is something we've been wanting to address. I like the approach.

**Comment by ahmetmeleq** [comment]
Looks all good now, waiting for tests to approve

**Review by ryannikolaidis — APPROVED** [review]
lgtm! thanks for the clean up here.

**Review by ahmetmeleq — APPROVED** [review]
Destination tests passed as well; LGTM, thanks for everything here.
